src/Swift.Runtime/src/Swift/SwiftArray.cs

```diff
@@ -0,0 +1,285 @@
+///    |   |   | var rawValue: IntPtr                                 |   |   |
+///    |   |   +------------------------------------------------------+   |   |
+///    |   +--------------------------------------------------------------+   |
+///    +-----------------------------------------------------------------------+
+/// </summary>
+[StructLayout(LayoutKind.Sequential, Size = 8)]
+public struct ArrayBuffer
+{
+    public IntPtr storage;
+}
+
+/// <summary>
+/// Represents a Swift array.
+/// </summary>
+/// <typeparam name="Element">The element type contained in the array.</typeparam>
+public class SwiftArray<Element> : IDisposable, ISwiftObject
+{
+    static nuint _payloadSize = SwiftObjectHelper<SwiftArray<Element>>.GetTypeMetadata().Size;
+    static nuint _elementSize = ElementTypeMetadata.Size;
+
+    // Swift array is a value type and doesn't contain an IntPtr payload
+    private ArrayBuffer buffer;
+    public unsafe void Dispose()
+    {
+        if (buffer.storage != IntPtr.Zero)
+        {
+            Arc.Release(*(IntPtr*)buffer.storage);
+            buffer.storage = IntPtr.Zero;
+            GC.SuppressFinalize(this);
+        }
+    }
+
+    unsafe ~SwiftArray()
+    {
+        Arc.Release(*(IntPtr*)buffer.storage);
+        buffer.storage = IntPtr.Zero;
+    }
+    public static nuint PayloadSize => _payloadSize;
+    public ArrayBuffer Payload => buffer;
+    public static nuint ElementSize => _elementSize;
+    static TypeMetadata ISwiftObject.GetTypeMetadata()
+    {
+        return TypeMetadata.Cache.GetOrAdd(typeof(SwiftArray<Element>), _ => SwiftArrayPInvokes.PInvoke_getMetadata(TypeMetadataRequest.Complete, ElementTypeMetadata));
+    }
+
+    static TypeMetadata ElementTypeMetadata
+    {
+        get => TypeMetadata.GetTypeMetadataOrThrow<Element>();
+    }
+
+    static ISwiftObject ISwiftObject.NewFromPayload(SwiftHandle handle)
+    {
+        return new SwiftArray<Element>(handle);
+    }
+
+    IntPtr ISwiftObject.MarshalToSwift(IntPtr swiftDest)
+    {
+        var metadata = SwiftObjectHelper<SwiftArray<Element>>.GetTypeMetadata();
+        unsafe
+        {
+            fixed (void* _payloadPtr = &buffer)
+            {
+                metadata.ValueWitnessTable->InitializeWithCopy((void*)swiftDest, (void*)_payloadPtr, metadata);
+            }
+        }
+        return swiftDest;
+    }
+
+    /// <summary>
+    /// Gets the protocol conformance descriptor for the given type.
+    /// </summary>
+    /// <typeparam name="TProtocol"></typeparam>
+    /// <returns></returns>
+    static ProtocolConformanceDescriptor ISwiftObject.GetProtocolConformanceDescriptor<TProtocol>()
+        where TProtocol : class
+    {
+        return ProtocolConformanceDescriptor.Zero;
+    }
+
+    /// <summary>
+    /// Constructs a new SwiftArray from the given handle.
+    /// </summary>
+    unsafe SwiftArray(SwiftHandle handle)
+    {
+        this.buffer = *(ArrayBuffer*)(handle);
```

Comment:
Should this just copy the memory? What about arrays of reference types?

Reply:
I would remove this constructor as it goes against the expected behavior on Swift side. Even if payload value is copied, it would point to the same instance.

Reply:
Can't be removed, it is used in the `ISwiftObject.NewFromPayload`. It should perform a shallow copy of the `ArrayBuffer` struct.

Anyway, it doesn't have the same semantics as in Swift.

Reply:
Ok, maybe this is a future problem once we start supporting reference types. We might need to call `InitializeWithCopy` here

Reply:
> maybe this is a future problem once we start supporting reference types

Could you provide an example?

> We might need to call InitializeWithCopy here

Added to https://github.com/dotnet/runtimelab/issues/2930.

Reply:
I don't have anything specific in my mind yet. I was think about having an array of reference types. If we just make a shallow copy of the array the ref count of those references will not be updated.